### DISCUSSION TITLE
PocketBaseUML

### DISCUSSION THREAD

**Discussion by icflorescu** [pocketbase/pocketbase#2090] [Show and tell]
[PocketBaseUML](https://pocketbase-uml.github.io/) is a free, open-source UML diagram generator/viewer for PocketBase.

You can use it online at [pocketbase-uml.github.io](https://pocketbase-uml.github.io/) to connect to a HTTPS PocketBase server, or run it as an `npm` package on your machine to connect to a local PocketBase instance through HTTP.

You can also import a `collections.json` file exported from PocketBase to view it as an UML diagram.

[![PocketBaseUML](https://user-images.githubusercontent.com/581999/225126356-d074ceb2-5139-419b-b93d-c4a62e3e8585.png)](https://pocketbase-uml.github.io/)

PocketBase comes with a fantastic UI already (better than Supabase or Firebase, IMO), but I've always missed the ability to actually SEE the data architecture in a graphical form, easy to parse in a quick glance.

The project is still in the early stage, there are still potential new features to be added and needs to be kept in sync with new upcoming PocketBase features.

Please have a look at it, give it a whirl and let me know in the comments if you find it useful or interesting!

**Comment by icflorescu** [comment]
It should be obvious and I already mentioned it in the initial post, but please be aware that, due to security browser restrictions,  **the online version at [pocketbase-uml.github.io](https://pocketbase-uml.github.io/) can only connect to HTTPS servers**.

**If you want to access a local instance running through HTTP, you must run PocketBaseUML locally** like so:
```sh
npx pocketbase-uml@latest
```

**Comment by dschissler** [comment]
@ganigeorgiev This project seems like a compelling reason for the admin ui to have plugin capability.

**Comment by dschissler** [comment]
@icflorescu Have you considered embedding the Svelte dist resources into a Golang package that can be mounted with Echo / Pocketbase?  That would allow this project to be installable with only a few lines of code.

**Reply by jellypeep** [reply]
Nice project! I am getting the following error when trying to import collecitons from a Pocketbase server 0.13.4:

`The provided code is either an invalid JSON or comes from a newer server version and PocketBaseUML doesn't know how to draw its diagram.`

**Reply by icflorescu** [reply]
@dschissler Sounds interesting, but I have no prior experience with Go yet. Something to keep in mind, though ;-)

**Reply by icflorescu** [reply]
@jellypeep Interesting. I'm testing with `0.13.4` and can't replicate this, but if you could send me the `collections.json` you're trying to import I might discover an edge case I haven't thought about and I'm sure I can address the problem.
It may be a good idea to open an [issue in the repo](https://github.com/pocketbase-uml/pocketbase-uml.github.io/issues) in order not to to pollute this Pocketbase discussion thread.
Thanks for testing it out!

**Reply by icflorescu** [reply]
@jellypeep The bug has been successfully identified - it was happening when collection names contained underscores.
The latest version fixes it.

**Reply by margani** [reply]
The error that @jellypeep sent is still happening. The exported collection uses name `schema` instead of `fields`. I am using `PocketBase v0.22.34`.

**Reply by margani** [reply]
I fixed it in https://github.com/pocketbase-uml/pocketbase-uml.github.io/pull/9

**Comment by hungcrush** [comment]
@icflorescu Do you think we can generate a docs like swagger ?

**Comment by icflorescu** [comment]
@hungcrush It definitely could be done. Not a trivial amount of work, though, and probably should be the focus of a different project.

And I have to agree with @dschissler, having some sort of a plugin system would be fantastic, as it would enable lots of people to come up with their own crazy ideas, ending up with a vast ecosystem of plugins that would make a such a great project even better.

I'm curious to hear @ganigeorgiev's opinion on this idea too...

**Reply by dschissler** [reply]
@icflorescu It would be embedded like [pocketbase admin ui](https://github.com/pocketbase/pocketbase/tree/master/ui) does it.  Maybe I'll try to refactor your code someday if you're not able.

@ganigeorgiev How do you develop admin ui?  Are you able to use vite developer server while building it or do you have to continually rebuild the dist source?  I can't seem to figure that out.

**Reply by ganigeorgiev** [reply]
The Admin UI at the moment is not extendable (_this eventually will be handled as part of https://github.com/pocketbase/pocketbase/issues/898_).

> Are you able to use vite developer server while building it or do you have to continually rebuild the dist source?

You can run `npm run dev`. This will start a dev server on `localhost:3000` but you'll also need the api on `:8090`.
For more details, you can also check https://github.com/pocketbase/pocketbase/blob/master/CONTRIBUTING.md#making-changes-in-the-admin-ui
